src/io/delimited.cr

```diff
@@ -0,0 +1,124 @@
+module IO
+  # An IO that wraps another IO, and only reads up to the beginning of a
+  # specified delimiter.
+  #
+  # This is useful for exposing part of an underlying stream to a client.
+  #
+  # ```
+  # io = MemoryIO.new "abc||123"
+  # delimited = IO::Delimited.new(io, read_delimiter: "||")
+  #
+  # delimited.gets_to_end # => "abc"
+  # delimited.gets_to_end # => nil
+  # io.gets_to_end        # => "123"
+  # ```
+  class Delimited
+    include IO
+
+    # If `sync_close` is true, closing this IO will close the underlying IO.
+    property? sync_close
+
+    getter read_delimiter
+    getter? closed : Bool
+
+    @delimiter_buffer : Bytes
+    @active_delimiter_buffer : Bytes?
```

Comment:
Line 42 and 43 kinda serve this purpose.

`active_delimiter_buffer` is simply the slice containing the part of `delimiter_buffer` that is actually in use and not garbage bytes.